### PR TITLE
[Pipetools/coverage] Make pipes work on windows + add tests

### PR DESCRIPTION
**Pull request by gpotter2**
This PR:
- re-use some of my work on `automaton.py` to replace `select.select` in `pipetools.py`
- add an implementation for Windows (there where some `xterm` calls ==> adapted the functions to work with powershell)

@phil777 You might want to have a look at this, as you did https://github.com/secdev/scapy/pull/620

### CODE REVIEW

**Comment by gpotter2** [comment]
@guedou Hello, could you review this one please ? It brings quite a lot of coverage

**Comment by gpotter2** [comment]
@p-l- Hey, could you review this one please ? It brings 0.85% of coverage 😃

File: scapy/pipetool.py [hunk 1]
```diff
@@ -474,26 +497,51 @@ def __init__(self, name=None, keepterm=True, newlines=True, openearly=True):
...
-                cmd.append("-hold")
-            cmd.extend(["-e", "cat 0<&%i" % self.__r])
-            self.__p = subprocess.Popen(cmd)
-            os.close(self.__r)
+            if WINDOWS:
+                self.__f = get_temp_file()
+                self.name = "Scapy" if self.name is None else self.name
+                # Start a powershell in a new window and print the PID
+                cmd = "$app = Start-Process PowerShell -ArgumentList '-command &{$host.ui.RawUI.WindowTitle=\\\"%s\\\";Get-Content \\\"%s\\\" -wait}' -passthru; echo $app.Id" % (self.name, self.__f.replace("\\", "\\\\"))
+                print([conf.prog.powershell, cmd])
```
Comment (p-l-) [inline scapy/pipetool.py]: Is this a forgotten debug `print()`?

File: scapy/pipetool.py [hunk 2]
```diff
@@ -474,26 +497,51 @@ def __init__(self, name=None, keepterm=True, newlines=True, openearly=True):
...
-            if self.keepterm:
-                cmd.append("-hold")
-            cmd.extend(["-e", "cat 0<&%i" % self.__r])
-            self.__p = subprocess.Popen(cmd)
-            os.close(self.__r)
+            if WINDOWS:
+                self.__f = get_temp_file()
+                self.name = "Scapy" if self.name is None else self.name
+                # Start a powershell in a new window and print the PID
+                cmd = "$app = Start-Process PowerShell -ArgumentList '-command &{$host.ui.RawUI.WindowTitle=\\\"%s\\\";Get-Content \\\"%s\\\" -wait}' -passthru; echo $app.Id" % (self.name, self.__f.replace("\\", "\\\\"))
```
Comment (p-l-) [inline scapy/pipetool.py]: Do we **have** to use powershell here?

File: scapy/pipetool.py [hunk 3]
```diff
@@ -474,26 +497,51 @@ def __init__(self, name=None, keepterm=True, newlines=True, openearly=True):
...
+                cmd.extend(["-e", "cat 0<&%i" % self.__r])
+                self.__p = subprocess.Popen(cmd, shell=True, executable="/bin/bash")
+                os.close(self.__r)
     def stop(self):
         if not self.keepterm:
             self.opened = False
-            os.close(self.__w)
-            self.__p.kill()
-            self.__p.wait()
+            if not WINDOWS:
```
Comment (p-l-) [inline scapy/pipetool.py]: Can you use a "positive" test here (`if WINDOWS:`)?

File: scapy/pipetool.py (hunk 2 above)
Comment (gpotter2) [inline scapy/pipetool.py]: The goal of the program was to start a new shell on the computer to print some stuff. To emulate `tail -f`, I use Get-Content, only available in Powershell.

File: scapy/pipetool.py (hunk 1 above)
Comment (gpotter2) [inline scapy/pipetool.py]: Yes...

File: scapy/pipetool.py [hunk 4]
```diff
@@ -474,26 +494,51 @@ def __init__(self, name=None, keepterm=True, newlines=True, openearly=True):
...
-            self.__r,self.__w = os.pipe()
-            cmd = ["xterm"]
-            if self.name is not None:
-                cmd.extend(["-title",self.name])
-            if self.keepterm:
-                cmd.append("-hold")
-            cmd.extend(["-e", "cat 0<&%i" % self.__r])
-            self.__p = subprocess.Popen(cmd)
-            os.close(self.__r)
+            if WINDOWS:
```
Comment (gpotter2) [inline scapy/pipetool.py]: Ok. But i would like to be able to call the functions separatly

File: scapy/pipetool.py [hunk 5]
```diff
@@ -55,6 +60,20 @@ def f(*args, **kargs):
                 return f
         raise AttributeError(attr)
 
+    def checkRecv(self):
```
Comment (gpotter2) [inline scapy/pipetool.py]: Sure

**Comment by guedou** [comment]
I thinkt that it is ready to be merged.

@p-l- what is your call ?

**Comment by gpotter2** [comment]
(I just rebased to fix the conflicts)

**Pull request merged**